File: backend/app/schemas/complaint.py
"""
Pydantic schemas for Complaint API request/response validation.

All API payloads and responses for the complaint feature flow through these.
"""

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# GPS / Location schemas
# ---------------------------------------------------------------------------

class GPSLocation(BaseModel):
    """Browser-captured GPS coordinates."""
    latitude: float = Field(..., description="Latitude in decimal degrees")
    longitude: float = Field(..., description="Longitude in decimal degrees")
    accuracy: Optional[float] = Field(None, description="Accuracy in meters")
    altitude: Optional[float] = Field(None, description="Altitude in meters")
    speed: Optional[float] = Field(None, description="Speed in m/s")
    heading: Optional[float] = Field(None, description="Heading in degrees")
    timestamp: Optional[datetime] = Field(None, description="GPS capture timestamp")


class GeoAddress(BaseModel):
    """Reverse-geocoded address components."""
    village: Optional[str] = None
    ward: Optional[str] = None
    taluka: Optional[str] = None
    district: Optional[str] = None
    state: Optional[str] = None
    pincode: Optional[str] = None
    assembly_constituency: Optional[str] = None
    lok_sabha_constituency: Optional[str] = None
    nearest_landmark: Optional[str] = None
    confidence: Optional[float] = Field(None, ge=0.0, le=1.0)
    raw_display: Optional[str] = None


# ---------------------------------------------------------------------------
# Evidence schemas
# ---------------------------------------------------------------------------

class EvidenceUpload(BaseModel):
    """Evidence attached to a complaint."""
    image_urls: List[str] = Field(default_factory=list)
    voice_url: Optional[str] = None
    video_url: Optional[str] = None
    text_description: Optional[str] = None


class EvidenceScore(BaseModel):
    """Calculated evidence score breakdown."""
    total: float = Field(..., ge=0, le=100, description="Overall evidence score 0-100")
    gps_present: bool = False
    photo_present: bool = False
    voice_present: bool = False
    image_metadata_valid: bool = False
    multiple_reports_nearby: bool = False
    duplicate_radius_match: bool = False
    location_accuracy_score: float = 0.0
    timestamp_valid: bool = False
    ai_confidence_score: float = 0.0
    breakdown: Dict[str, float] = Field(default_factory=dict)


# ---------------------------------------------------------------------------
# Duplicate / Cluster schemas
# ---------------------------------------------------------------------------

class DuplicateCheckResult(BaseModel):
    """Result of duplicate radius search."""
    is_duplicate: bool = False
    cluster_id: Optional[str] = None
    cluster_uid: Optional[str] = None
    existing_report_count: int = 0
    nearest_distance_meters: Optional[float] = None
    merged_into_cluster: bool = False


class IssueClusterInfo(BaseModel):
    """Issue cluster summary for display."""
    id: str
    cluster_uid: str
    report_count: int
    category: str
    department: Optional[str] = None
    centroid_lat: float
    centroid_lng: float
    radius_meters: float
    severity: str
    village: Optional[str] = None
    ward: Optional[str] = None


# ---------------------------------------------------------------------------
# AI Preview schemas
# ---------------------------------------------------------------------------

class AIPreview(BaseModel):
    """AI-generated preview shown before complaint submission."""
    detected_department: Optional[str] = None
    detected_sector: Optional[str] = None
    detected_category: Optional[str] = None
    detected_location: Optional[GeoAddress] = None
    gps_accuracy: Optional[float] = None
    evidence_score: Optional[EvidenceScore] = None
    duplicate_probability: float = 0.0
    priority_prediction: float = 0.0
    estimated_resolution_days: Optional[int] = None
    ai_confidence: float = 0.0
    similar_complaints_nearby: int = 0


# ---------------------------------------------------------------------------
# Complaint Create / Request
# ---------------------------------------------------------------------------

class ComplaintCreateRequest(BaseModel):
    """Full complaint submission payload from citizen."""
    title: str = Field(..., min_length=3, max_length=300)
    description: Optional[str] = None
    category: str = Field(..., description="Complaint category")
    sector: Optional[str] = None

    # GPS
    gps: Optional[GPSLocation] = None

    # Geocoded address (populated by backend if not provided)
    address: Optional[GeoAddress] = None

    # Evidence
    evidence: Optional[EvidenceUpload] = None

    # Citizen info
    citizen_id: Optional[str] = None
    citizen_name: Optional[str] = None

    # Optional manual corrections
    manual_ward: Optional[str] = None
    manual_village: Optional[str] = None


# ---------------------------------------------------------------------------
# Complaint Response
# ---------------------------------------------------------------------------

class ComplaintResponse(BaseModel):
    """Full complaint response with all geo, evidence, and AI fields."""
    id: str
    complaint_uid: str
    title: str
    description: Optional[str] = None

    category: str
    sector: Optional[str] = None
    department: Optional[str] = None
    severity: str
    status: str

    # GPS
    gps_latitude: Optional[float] = None
    gps_longitude: Optional[float] = None
    gps_accuracy: Optional[float] = None
    gps_altitude: Optional[float] = None
    gps_speed: Optional[float] = None
    gps_heading: Optional[float] = None

    # Geocoded address
    village: Optional[str] = None
    ward: Optional[str] = None
    taluka: Optional[str] = None
    district: Optional[str] = None
    state: Optional[str] = None
    pincode: Optional[str] = None
    assembly_constituency: Optional[str] = None
    lok_sabha_constituency: Optional[str] = None
    nearest_landmark: Optional[str] = None

    # Verification
    verification_status: str
    verification_confidence: float
    evidence_score: float

    # Duplicate
    duplicate_probability: float
    cluster_id: Optional[str] = None
    cluster_uid: Optional[str] = None
    duplicate_count: int

    # AI
    ai_detected_category: Optional[str] = None
    ai_detected_department: Optional[str] = None
    ai_detected_sector: Optional[str] = None
    ai_confidence: float
    priority_prediction: float
    estimated_resolution_days: Optional[int] = None

    # Evidence
    images: Optional[List[str]] = None
    voice_url: Optional[str] = None

    # Citizen
    citizen_id: Optional[str] = None
    citizen_name: Optional[str] = None

    # Heatmap
    heatmap_key: Optional[str] = None

    # Timestamps
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# API wrapper response
# ---------------------------------------------------------------------------

class ComplaintSubmitResponse(BaseModel):
    """Response after submitting a complaint – includes AI preview data."""
    success: bool = True
    complaint: ComplaintResponse
    ai_preview: Optional[AIPreview] = None
    duplicate_info: Optional[DuplicateCheckResult] = None
    cluster_info: Optional[IssueClusterInfo] = None
    message: str = "Complaint submitted and verified successfully"


class ComplaintListResponse(BaseModel):
    """Paginated list of complaints."""
    complaints: List[ComplaintResponse]
    total: int
    page: int
    page_size: int


class ComplaintStatsResponse(BaseModel):
    """Complaint statistics for dashboard."""
    total_complaints: int = 0
    verified_complaints: int = 0
    pending_complaints: int = 0
    resolved_complaints: int = 0
    avg_evidence_score: float = 0.0
    total_clusters: int = 0

File: frontend/src/types/complaint.ts
/**
 * Complaint types for the AI Geo-Verified Complaint Engine.
 */

// ─── GPS Location ─────────────────────────────────────────────────────
export interface GPSLocation {
  latitude: number;
  longitude: number;
  accuracy: number | null;
  altitude: number | null;
  speed: number | null;
  heading: number | null;
  timestamp: string | null;
}

// ─── Geo Address ──────────────────────────────────────────────────────
export interface GeoAddress {
  village: string | null;
  ward: string | null;
  taluka: string | null;
  district: string | null;
  state: string | null;
  pincode: string | null;
  assembly_constituency: string | null;
  lok_sabha_constituency: string | null;
  nearest_landmark: string | null;
  confidence: number | null;
  raw_display: string | null;
}

// ─── Evidence Score ───────────────────────────────────────────────────
export interface EvidenceScore {
  total: number;
  gps_present: boolean;
  photo_present: boolean;
  voice_present: boolean;
  image_metadata_valid: boolean;
  multiple_reports_nearby: boolean;
  duplicate_radius_match: boolean;
  location_accuracy_score: number;
  timestamp_valid: boolean;
  ai_confidence_score: number;
  breakdown: Record<string, number>;
}

// ─── Evidence Upload ──────────────────────────────────────────────────
export interface EvidenceUpload {
  image_urls: string[];
  voice_url: string | null;
  video_url: string | null;
  text_description: string | null;
}

// ─── Duplicate Check ──────────────────────────────────────────────────
export interface DuplicateCheckResult {
  is_duplicate: boolean;
  cluster_id: string | null;
  cluster_uid: string | null;
  existing_report_count: number;
  nearest_distance_meters: number | null;
  merged_into_cluster: boolean;
}

// ─── Issue Cluster ────────────────────────────────────────────────────
export interface IssueClusterInfo {
  id: string;
  cluster_uid: string;
  report_count: number;
  category: string;
  department: string | null;
  centroid_lat: number;
  centroid_lng: number;
  radius_meters: number;
  severity: string;
  village: string | null;
  ward: string | null;
}

// ─── AI Preview ───────────────────────────────────────────────────────
export interface AIPreview {
  detected_department: string | null;
  detected_sector: string | null;
  detected_category: string | null;
  detected_location: GeoAddress | null;
  gps_accuracy: number | null;
  evidence_score: EvidenceScore | null;
  duplicate_probability: number;
  priority_prediction: number;
  estimated_resolution_days: number | null;
  ai_confidence: number;
  similar_complaints_nearby: number;
}

// ─── Complaint ────────────────────────────────────────────────────────
export interface Complaint {
  id: string;
  complaint_uid: string;
  title: string;
  description: string | null;
  category: string;
  sector: string | null;
  department: string | null;
  severity: string;
  status: string;

  // GPS
  gps_latitude: number | null;
  gps_longitude: number | null;
  gps_accuracy: number | null;
  gps_altitude: number | null;
  gps_speed: number | null;
  gps_heading: number | null;

  // Address
  village: string | null;
  ward: string | null;
  taluka: string | null;
  district: string | null;
  state: string | null;
  pincode: string | null;
  assembly_constituency: string | null;
  lok_sabha_constituency: string | null;
  nearest_landmark: string | null;

  // Verification
  verification_status: string;
  verification_confidence: number;
  evidence_score: number;

  // Duplicate
  duplicate_probability: number;
  cluster_id: string | null;
  cluster_uid: string | null;
  duplicate_count: number;

  // AI
  ai_detected_category: string | null;
  ai_detected_department: string | null;
  ai_detected_sector: string | null;
  ai_confidence: number;
  priority_prediction: number;
  estimated_resolution_days: number | null;

  // Evidence
  images: string[] | null;
  voice_url: string | null;

  // Citizen
  citizen_id: string | null;
  citizen_name: string | null;

  // Heatmap
  heatmap_key: string | null;

  // Timestamps
  created_at: string;
  updated_at: string;
}

// ─── API Response Types ───────────────────────────────────────────────
export interface ComplaintSubmitResponse {
  success: boolean;
  complaint: Complaint;
  ai_preview: AIPreview | null;
  duplicate_info: DuplicateCheckResult | null;
  cluster_info: IssueClusterInfo | null;
  message: string;
}

export interface ComplaintListResponse {
  complaints: Complaint[];
  total: number;
  page: number;
  page_size: number;
}

export interface ComplaintStatsResponse {
  total_complaints: number;
  verified_complaints: number;
  pending_complaints: number;
  resolved_complaints: number;
  avg_evidence_score: number;
  total_clusters: number;
}

// ─── Request Types ────────────────────────────────────────────────────
export interface ComplaintCreateRequest {
  title: string;
  description: string | null;
  category: string;
  sector: string | null;
  gps: GPSLocation | null;
  address: GeoAddress | null;
  evidence: EvidenceUpload | null;
  citizen_id: string | null;
  citizen_name: string | null;
  manual_ward: string | null;
  manual_village: string | null;
  // Multilingual fields
  original_language?: string;
  original_text?: string;
  english_translation?: string;
  final_edited_text?: string;
}

// ─── GPS Permission State ─────────────────────────────────────────────
export type GPSPermissionState = "prompt" | "granted" | "denied" | "unavailable";

// ─── Complaint Category (extended) ────────────────────────────────────
export type ComplaintCategory =
  | "road"
  | "water"
  | "electricity"
  | "healthcare"
  | "education"
  | "sanitation"
  | "other";
